syntax = "proto3";

package raft;

message config {
    message server {
        uint32 id = 1;
        string host = 2;
        uint32 port = 3;
    }
    server self = 2;
    repeated server peers = 3;
}
message log_entry {
    uint64 term = 1;
}

message append_entries_request {
    uint64 term = 1;
    uint32 leader_id = 2;
    uint64 prev_log_index = 3;
    uint64 prev_log_term = 4;
    uint64 leader_commit = 5;
    repeated log_entry entries = 6;
}

message append_entries_response {
    uint64 term = 1;
    bool success = 2;
}

message vote_request {
    uint64 term = 1;
    uint32 candidate_id = 2;
    uint64 last_log_index = 3;
    uint64 last_log_term = 4;
}
message vote_response {
    uint64 term = 1;
    bool granted = 2;
}

message raft_message {
    enum type {
        NONE = 0;
        APPEND_ENTRIES = 1;
        VOTE = 2;
    }
    type discriminator = 1;

    append_entries_request append_entries_request = 2;
    append_entries_response append_entries_response = 3;
    vote_request vote_request = 4;
    vote_response vote_response = 5;
}
